packages/xstate-graph/src/graph.ts

```diff
@@ -393,64 +348,295 @@ export function toDirectedGraph(
+    SerializedState,
+    { state: TState; paths: Array<StatePath<TState, TEvent>> }
+  > = {};
+
+  function util(
+    fromState: TState,
+    toStateSerial: SerializedState,
+    event: TEvent | null
+  ) {
+    const fromStateSerial = serializeState(fromState, event);
+    visitCtx.vertices.add(fromStateSerial);
+
+    if (fromStateSerial === toStateSerial) {
+      if (!pathMap[toStateSerial]) {
+        pathMap[toStateSerial] = {
+          state: stateMap.get(toStateSerial)!,
+          paths: []
+        };
+      }
+
+      const toStatePlan = pathMap[toStateSerial];
+
+      const path2: StatePath<TState, TEvent> = {
+        state: fromState,
+        weight: path.length,
+        steps: [...path]
+      };
+
+      toStatePlan.paths.push(path2);
+    } else {
+      for (const serializedEvent of Object.keys(
+        adjacency[fromStateSerial].transitions
+      ) as SerializedEvent[]) {
+        const { state: nextState, event: subEvent } = adjacency[
+          fromStateSerial
+        ].transitions[serializedEvent];
+
+        if (!(serializedEvent in adjacency[fromStateSerial].transitions)) {
+          continue;
+        }
+
+        const nextStateSerial = serializeState(nextState, subEvent);
+        stateMap.set(nextStateSerial, nextState);
+
+        if (!visitCondition(nextState, subEvent, visitCtx)) {
+          visitCtx.edges.add(serializedEvent);
+          path.push({
+            state: stateMap.get(fromStateSerial)!,
+            event: subEvent
+          });
+          util(nextState, toStateSerial, subEvent);
+        }
+      }
+    }
+
+    path.pop();
+    visitCtx.vertices.delete(fromStateSerial);
+  }
+
+  const initialStateSerial = serializeState(initialState, null);
+  stateMap.set(initialStateSerial, initialState);
+
+  for (const nextStateSerial of Object.keys(adjacency) as SerializedState[]) {
+    util(initialState, nextStateSerial, null);
+  }
+
+  return Object.values(pathMap);
+}
+
+export function filterPlans<TState, TEvent extends EventObject>(
+  plans: Array<StatePlan<TState, TEvent>>,
+  predicate: (state: TState, plan: StatePlan<TState, TEvent>) => boolean
+): Array<StatePlan<TState, TEvent>> {
+  const filteredPlans = plans.filter((plan) => predicate(plan.state, plan));
+
+  return filteredPlans;
+}
+
+export function traverseSimplePathsTo<TState, TEvent extends EventObject>(
+  behavior: SimpleBehavior<TState, TEvent>,
+  predicate: (state: TState) => boolean,
+  options: TraversalOptions<TState, TEvent>
+): Array<StatePlan<TState, TEvent>> {
+  const resolvedOptions = resolveTraversalOptions(options);
+  const simplePlansMap = traverseSimplePlans(behavior, resolvedOptions);
+
+  return filterPlans(simplePlansMap, predicate);
+}
+
+export function traverseSimplePathsFromTo<TState, TEvent extends EventObject>(
+  behavior: SimpleBehavior<TState, TEvent>,
+  fromPredicate: (state: TState) => boolean,
+  toPredicate: (state: TState) => boolean,
+  options: TraversalOptions<TState, TEvent>
```

Comment:
maybe those could be merged in a single options param? 4 positional arguments is already a lot to memorize

Reply:
Nit-level: I'm happy not to change this.